rfcs/text/0003_handler_interface.md

```diff
@@ -0,0 +1,340 @@
+   their business logic just so they can access them within the context of a
+   handler.
+
+5. Functionality provided by services should not be arbitrarily used in
+   unconstrained execution such as in the plugin lifecycle hooks. For example,
+   it's appropriate for an Elasticsearch client to throw an error if it's used
+   inside an API route and Elasticsearch isn't available, however it's not
+   appropriate for a plugin to throw an error in their start function if
+   Elasticsearch is not available. If the ES client was only made available
+   within the handler context and not to the plugin's start contract at large,
+   then this isn't an issue we'll encounter in the first place.
+
+# Detailed design
+
+There are two parts to this proposal. The first is the handler interface
+itself, and the second is the interface that a service owner implements to make
+their handlers extensible.
+
+## Handler Context
+
+```ts
+interface Context {
+  core: unknown;
+  [contextName: string]: unknown;
+}
+
+type Handler = (context: Partial<Context>, ...args: unknown[]) => Promise<unknown>;
+```
+
+- `args` in this example is specific to the handler type, for instance in a
+  http route handler, this would include the incoming request object.
+- The context object is marked as `Partial<Context>` because the contexts
+  available will vary depending on which plugins are enabled.
+- This type is a convention, not a concrete type. The `core` key should have a
+  known interface that is declared in the service owner's specific Context type.
+
+## Registering new contexts
+
+```ts
+type ContextProvider<T extends keyof Context> = (...args: unknown[]) => Promise<Context[T]>;
+
+interface HandlerService {
+  registerContext<T extends keyof Context>(contextName: T, provider: ContextProvider<T>): void;
+}
+```
+
+- `args` in this example is specific to the handler type, for instance in a http
+  route handler, this would include the incoming request object. It would not
+  include the results from the other context providers in order to keep
+  providers from having dependencies on one another.
+- The `HandlerService` is defined as a literal interface in this document, but
+  in practice this interface is just a guide for the pattern of registering
+  context values. Certain services may have multiple different types of
+  handlers, so they may choose not to use the generic name `registerContext` in
+  favor of something more explicit.
+
+## Context creation
+
+Before a handler is executed, each registered context provider will be called
+with the given arguments to construct a context object for the handler. Each
+provider must return an object of the correct type. The return values of these
+providers is merged into a single object where each key of the object is the
+name of the context provider and the value is the return value of the provider.
+Key facts about context providers:
+
+- **Context providers cannot access context from other providers.** They should
```

Comment:
I think this is an issue. Consider any client that relies on the ES client, like an alerting client or saved object client. Every time you construct one of these clients for a context, you'd want to make sure you're using the latest Elasticsearch client, which itself would be made available via context. This issue compounds on itself if your client relies on something like saved objects, which then rely on the es client.

To support this, I propose making the order deterministic based on the order in which the providers were registered. Coupled with the intentional order of our plugin lifecycle execution which guarantees that dependencies execute first, this would ensure that a provider would have access to the context constructed by all of its dependencies.

Reply:
This is a good point. This also made me realize that we should make sure that we do not leak the existence of plugin contexts for dependencies you do not declare via the context object. The service owner for each handler will need to implement this mechanism, but it is a solvable problem.

RFC updated.